### PR TITLE
Fix deprecations, improve messaging

### PR DESCRIPTION
**Pull request by jakebailey**
As noted in #52975, things don't appear to be working as expected.

This PR makes it so that for every set of deprecated values, there is:

- The version in which the option became deprecated.
- The version in which the option has been removed.

If the current version is greater than or equal to the version in which the option was removed, then you get an error saying that the option has been removed and must be removed from `tsconfig.json`.

Otherwise, if the `ignoreDeprecation` option is _not_ greater than or equal to the version in which the option became deprecated, then we'll issue an error saying to set `ignoreDeprecation` to at least the version that allows there to be no errors.

Along the way:

- Make it so that we don't say "flag"; not all of these are flags.
- Make it so that if we are deprecating a specific value, we don't say "Flag ES3" or "Option ES3", but instead something like `Option 'target=ES3'`.

What changes in this PR in baselines is all messages, however, one test includes what happens if the version is 6.0; I think the new behavior is fine. There's nothing special about 6.0 compared to 5.5.

### CODE REVIEW

File: src/compiler/program.ts [hunk 1]
```diff
@@ -4320,97 +4322,115 @@ export function createProgram(rootNamesOrOptions: readonly string[] | CreateProg
         }
     }
 
-    function getVersionForDeprecationDiagnostics(reportInvalidIgnoreDeprecations: boolean) {
-        const version = typeScriptVersion || versionMajorMinor;
+    function getTypeScriptVersion(): [version: Version, versionString: string] {
```
Comment (jakebailey) [inline src/compiler/program.ts:4325]: This function is overcomplicated due to some old stuff I was doing. Will change.

**Comment by jakebailey** [comment]
This likely needs to be backported to 5.0, otherwise it will break if we introduce any other deprecations.

**Comment by jakebailey** [comment]
(but, I have a small refractor to apply before doing so)

**Review by sheetalkamat — APPROVED** [review]

File: src/compiler/program.ts [hunk 2]
```diff
@@ -4320,97 +4322,110 @@ export function createProgram(rootNamesOrOptions: readonly string[] | CreateProg
...
-        const version = typeScriptVersion || versionMajorMinor;
+    function getIgnoreDeprecationsVersion(): Version {
         const ignoreDeprecations = options.ignoreDeprecations;
         if (ignoreDeprecations) {
-            if (ignoreDeprecations === DeprecationVersion.v5_0 && version.startsWith("5.")) {
-                return;
-            }
-            else if (reportInvalidIgnoreDeprecations) {
-                createOptionValueDiagnostic("ignoreDeprecations", Diagnostics.Invalid_value_for_ignoreDeprecations);
+            const parsed = Version.tryParse(ignoreDeprecations);
```
Comment (jakebailey) [inline src/compiler/program.ts:4328]: We talked about this and it seems like the intent is that we are not going to deprecate _anything_ between TS 5.0 and TS 6.0, so this is overkill, and we can just write `=== "5.0"` and error.

**Review by sheetalkamat — APPROVED** [review]

File: src/compiler/program.ts [hunk 3]
```diff
@@ -4320,97 +4322,112 @@ export function createProgram(rootNamesOrOptions: readonly string[] | CreateProg
...
+                createDeprecatedDiagnostic("suppressImplicitAnyIndexErrors");
+            }
+            if (options.noStrictGenericChecks) {
+                createDeprecatedDiagnostic("noStrictGenericChecks");
+            }
+            if (options.charset) {
+                createDeprecatedDiagnostic("charset");
+            }
+            if (options.out) {
+                createDeprecatedDiagnostic("out");
```
Comment (DanielRosenwasser) [inline src/compiler/program.ts:4410]:
```suggestion
                createDeprecatedDiagnostic("out", /*value*/ undefined, "outFile");
```

**Review by DanielRosenwasser — APPROVED** [review]

**Review by RyanCavanaugh — APPROVED** [review]

**Comment by jakebailey** [comment]
@typescript-bot cherry-pick this to release-5.0

**Comment by jakebailey** [comment]
figures

**Pull request merged**